flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -395,6 +417,9 @@ public final void invoke() throws Exception {
 				// make sure no new timers can come
 				timerService.quiesce();
 
+				// let mailbox execution reject all new letters from this point
+				taskMailboxExecutor.shutdown();
```

Comment:
Is it possible that there are still some letters in the mailbox queue that would require to be executed (including future cases to host timer service and checkpoints via mailbox)?
Can `StreamTask::closeAllOperators()` still need the mailbox loop (e.g. for timers)?

My impression, that the old `StreamTask::run()` method (prior #8442 PR) finished running, when the source has ended. Other threads could still do valid operations
So the mailbox loop run is not the same as `StreamTask::run()`, but should span until `StreamTask.isRunning` is `true`.

Otherwise, we would need to change this code again (the `StreamTask` and the loop in `SourceStreamTask`) to support timers in compatible way.

Reply:
Currently I don't see any concrete case that would need this. Timers, checkpoints, operators, etc should all be completed at this point. If an operator needs to ensure that something completes first, it can do so in its `close` method - like the Async operator will do for example. For as long as there is no real use case I would strictly avoid running anything here so that the shutdown can complete asap.